rust/phoenix-channel/src/login_url.rs

```diff
@@ -0,0 +1,224 @@
+            firezone_token,
+            "client",
+            Some(public_key),
+            Some(external_id),
+            Some(device_name),
+            None,
+            None,
+        )?;
+
+        Ok(LoginUrl {
+            host: parse_host(&url)?,
+            url,
+        })
+    }
+
+    pub fn gateway<E>(
+        url: impl TryInto<Url, Error = E>,
+        firezone_token: &SecretString,
+        device_id: String,
+        device_name: Option<String>,
+        public_key: [u8; 32],
+    ) -> std::result::Result<Self, LoginUrlError<E>> {
+        let external_id = hex::encode(sha2::Sha256::digest(device_id));
+        let device_name = device_name
+            .or(get_host_name())
+            .unwrap_or_else(|| Uuid::new_v4().to_string());
+
+        let url = get_websocket_path(
+            url.try_into().map_err(LoginUrlError::InvalidUrl)?,
+            firezone_token,
+            "gateway",
+            Some(public_key),
+            Some(external_id),
+            Some(device_name),
+            None,
+            None,
+        )?;
+
+        Ok(LoginUrl {
+            host: parse_host(&url)?,
+            url,
+        })
+    }
+
+    pub fn relay<E>(
+        url: impl TryInto<Url, Error = E>,
+        firezone_token: &SecretString,
+        device_name: Option<String>,
+        ipv4_address: Option<Ipv4Addr>,
+        ipv6_address: Option<Ipv6Addr>,
+    ) -> std::result::Result<Self, LoginUrlError<E>> {
+        let url = get_websocket_path(
+            url.try_into().map_err(LoginUrlError::InvalidUrl)?,
+            firezone_token,
+            "relay",
+            None,
+            None,
+            device_name,
+            ipv4_address,
+            ipv6_address,
+        )?;
+
+        Ok(LoginUrl {
+            host: parse_host(&url)?,
+            url,
+        })
+    }
+
+    // TODO: Only temporarily public until we delete other phoenix-channel impl.
+    pub fn inner(&self) -> &Url {
+        &self.url
+    }
+
+    // TODO: Only temporarily public until we delete other phoenix-channel impl.
+    pub fn host(&self) -> &str {
+        &self.host
+    }
+}
+
+fn parse_host<E>(url: &Url) -> Result<String, LoginUrlError<E>> {
+    let host = url.host_str().ok_or(LoginUrlError::MissingHost)?;
+
+    Ok(match url.port() {
+        Some(p) => format!("{host}:{p}"),
+        None => host.to_owned(),
+    })
+}
+
+#[derive(Debug, thiserror::Error)]
+pub enum LoginUrlError<E> {
+    #[error("invalid scheme `{0}`; only http(s) and ws(s) are allowed")]
+    InvalidUrlScheme(String),
+    #[error("failed to parse URL: {0}")]
+    InvalidUrl(E),
+    #[error("the url is missing a host")]
+    MissingHost,
+}
+
+#[cfg(not(target_os = "windows"))]
+fn get_host_name() -> Option<String> {
+    let mut buf = [0; HOST_NAME_MAX];
+    // SAFETY: we just allocated a buffer with that size
+    if unsafe { libc::gethostname(buf.as_mut_ptr() as *mut _, HOST_NAME_MAX) } != 0 {
```

Comment:
What was the reason to not use the `hostname` crate on all platforms? internally they make the same call https://github.com/svartalf/hostname/blob/master/src/nix.rs#L25

Although, they don't enforce the 256 byte limit that we do. They allow higher if the OS does, it looks like.

Reply:
I don't know, I am just moving code around here :)